### ISSUE TITLE
PLAYING A TRACK

### ISSUE THREAD

**Issue by KeesKemper** [bazwilliams/openhomedevice#17]
Executing the following code:

from openhomedevice.Device import Device
location="http://192.168.0.140:33673/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/desc.xml"
openhomeDevice = Device(location)
trackDetails = {}
trackDetails["uri"] = "http://opml.radiotime.com/Tune.ashx?id=s122119"
trackDetails["title"] = 'Linn Radio (Eclectic Music)'
trackDetails["albumArtwork"] = 'http://cdn-radiotime-logos.tunein.com/s122119q.png'
openhomeDevice.PlayMedia(trackDetails)

Gives the following error:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/pi/.local/lib/python2.7/site-packages/openhomedevice/Device.py", line 83, in PlayMedia
    soapRequest(service.ControlUrl(), service.Type(), "SetChannel", channelValueString)
AttributeError: 'NoneType' object has no attribute 'ControlUrl'

What am I doing wrong?

**Comment by bazwilliams** [comment]
Can you provide a copy of the description at http://192.168.0.140:33673/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/desc.xml please?

**Comment by KeesKemper** [comment]
This XML file does not appear to have any style information associated with it. The document tree is shown below.
<root xmlns="urn:schemas-upnp-org:device-1-0">
<specVersion>
<major>1</major>
<minor>0</minor>
</specVersion>
<device>
<deviceType>urn:linn-co-uk:device:Source:1</deviceType>
<UDN>uuid:509a3dc9-d32b-30a1-ffff-ffff8842af55</UDN>
<friendlyName>VoltaStream-63 (OpenHome)</friendlyName>
<manufacturer>Ivo Clarysse, Henner Zeller (Bubblesoft)</manufacturer>
<manufacturerURL>
http://forum.xda-developers.com/showthread.php?t=1118891
</manufacturerURL>
<modelDescription>gmediarender 0.0.7-git (OpenHome)</modelDescription>
<modelName>gmediarender (OpenHome)</modelName>
<modelNumber>0.0.7-git (0.9-update37)</modelNumber>
<modelURL>http://github.com/hzeller/gmrender-resurrect</modelURL>
<iconList>
<icon>
<mimetype>image/png</mimetype>
<width>64</width>
<height>64</height>
<depth>24</depth>
<url>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/upnp/grender-64x64.png
</url>
</icon>
<icon>
<mimetype>image/png</mimetype>
<width>128</width>
<height>128</height>
<depth>24</depth>
<url>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/upnp/grender-128x128.png
</url>
</icon>
</iconList>
<serviceList>
<service>
<serviceType>urn:av-openhome-org:service:Product:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Product</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Volume:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Volume</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Time:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Time</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Info:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Info</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Playlist:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Playlist</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Credentials:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Credentials</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/desc.xml
</SCPDURL>
</service>
</serviceList>
</device>
</root>

**Comment by KeesKemper** [comment]
This XML file does not appear to have any style information associated with it. The document tree is shown below.
<root xmlns="urn:schemas-upnp-org:device-1-0">
<specVersion>
<major>1</major>
<minor>0</minor>
</specVersion>
<device>
<deviceType>urn:linn-co-uk:device:Source:1</deviceType>
<UDN>uuid:509a3dc9-d32b-30a1-ffff-ffff8842af55</UDN>
<friendlyName>VoltaStream-63 (OpenHome)</friendlyName>
<manufacturer>Ivo Clarysse, Henner Zeller (Bubblesoft)</manufacturer>
<manufacturerURL>
http://forum.xda-developers.com/showthread.php?t=1118891
</manufacturerURL>
<modelDescription>gmediarender 0.0.7-git (OpenHome)</modelDescription>
<modelName>gmediarender (OpenHome)</modelName>
<modelNumber>0.0.7-git (0.9-update37)</modelNumber>
<modelURL>http://github.com/hzeller/gmrender-resurrect</modelURL>
<iconList>
<icon>
<mimetype>image/png</mimetype>
<width>64</width>
<height>64</height>
<depth>24</depth>
<url>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/upnp/grender-64x64.png
</url>
</icon>
<icon>
<mimetype>image/png</mimetype>
<width>128</width>
<height>128</height>
<depth>24</depth>
<url>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/upnp/grender-128x128.png
</url>
</icon>
</iconList>
<serviceList>
<service>
<serviceType>urn:av-openhome-org:service:Product:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Product</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Volume:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Volume</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Time:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Time</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Info:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Info</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Playlist:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Playlist</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/desc.xml
</SCPDURL>
</service>
<service>
<serviceType>urn:av-openhome-org:service:Credentials:1</serviceType>
<serviceId>urn:av-openhome-org:serviceId:Credentials</serviceId>
<controlURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/action
</controlURL>
<eventSubURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/event
</eventSubURL>
<SCPDURL>
/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/desc.xml
</SCPDURL>
</service>
</serviceList>
</device>
</root>

**Comment by KeesKemper** [comment]
<?xml version="1.0" encoding="utf-8" standalone="yes"?><root xmlns="urn:schemas-upnp-org:device-1-0"><specVersion><major>1</major><minor>0</minor></specVersion><device><deviceType>urn:linn-co-uk:device:Source:1</deviceType><UDN>uuid:509a3dc9-d32b-30a1-ffff-ffff8842af55</UDN><friendlyName>VoltaStream-63 (OpenHome)</friendlyName><manufacturer>Ivo Clarysse, Henner Zeller (Bubblesoft)</manufacturer><manufacturerURL>http://forum.xda-developers.com/showthread.php?t=1118891</manufacturerURL><modelDescription>gmediarender 0.0.7-git (OpenHome)</modelDescription><modelName>gmediarender (OpenHome)</modelName><modelNumber>0.0.7-git (0.9-update37)</modelNumber><modelURL>http://github.com/hzeller/gmrender-resurrect</modelURL><iconList><icon><mimetype>image/png</mimetype><width>64</width><height>64</height><depth>24</depth><url>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/upnp/grender-64x64.png</url></icon><icon><mimetype>image/png</mimetype><width>128</width><height>128</height><depth>24</depth><url>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/upnp/grender-128x128.png</url></icon></iconList><serviceList><service><serviceType>urn:av-openhome-org:service:Product:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Product</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Volume:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Volume</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Time:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Time</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Info:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Info</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Playlist:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Playlist</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Credentials:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Credentials</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/desc.xml</SCPDURL></service></serviceList></device></root>

**Comment by KeesKemper** [comment]
<?xml version="1.0" encoding="utf-8" standalone="yes"?><root xmlns="urn:schemas-upnp-org:device-1-0"><specVersion><major>1</major><minor>0</minor></specVersion><device><deviceType>urn:linn-co-uk:device:Source:1</deviceType><UDN>uuid:509a3dc9-d32b-30a1-ffff-ffff8842af55</UDN><friendlyName>VoltaStream-63 (OpenHome)</friendlyName><manufacturer>Ivo Clarysse, Henner Zeller (Bubblesoft)</manufacturer><manufacturerURL>http://forum.xda-developers.com/showthread.php?t=1118891</manufacturerURL><modelDescription>gmediarender 0.0.7-git (OpenHome)</modelDescription><modelName>gmediarender (OpenHome)</modelName><modelNumber>0.0.7-git (0.9-update37)</modelNumber><modelURL>http://github.com/hzeller/gmrender-resurrect</modelURL><iconList><icon><mimetype>image/png</mimetype><width>64</width><height>64</height><depth>24</depth><url>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/upnp/grender-64x64.png</url></icon><icon><mimetype>image/png</mimetype><width>128</width><height>128</height><depth>24</depth><url>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/upnp/grender-128x128.png</url></icon></iconList><serviceList><service><serviceType>urn:av-openhome-org:service:Product:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Product</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Product/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Volume:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Volume</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Volume/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Time:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Time</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Time/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Info:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Info</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Info/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Playlist:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Playlist</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Playlist/desc.xml</SCPDURL></service><service><serviceType>urn:av-openhome-org:service:Credentials:1</serviceType><serviceId>urn:av-openhome-org:serviceId:Credentials</serviceId><controlURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/action</controlURL><eventSubURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/event</eventSubURL><SCPDURL>/dev/509a3dc9-d32b-30a1-ffff-ffff8842af55/svc/av-openhome-org/Credentials/desc.xml</SCPDURL></service></serviceList></device></root>

**Comment by bazwilliams** [comment]
If you're opening this in a web browser, would it be possible to right click and 'save as' the results of that URL? I should be able to figure it out - but would be easier with the original XML file.

**Comment by KeesKemper** [comment]
[desc.zip](https://github.com/bazwilliams/openhomedevice/files/4544919/desc.zip)

**Comment by bazwilliams** [comment]
I've managed to replicate this in my test suite and I've realised the problem lies with a lack of 'radio' service on the software device. 

The `PlayMedia` call is designed to play the audio item using the Radio source so as not to disrupt any playlist items currently enqueued. (See 1)

Enqueing the audioitem on the playlist requires thought into _where_ in the playlist the audioitem is added (at the beginning, at the end, or immediately after the currently playing track). Further thought is required if the playlist happened to already be full. Yet more questions arise if the playlist should be deleted and replaced with incoming audioItems. To avoid all this discussion, I decided to just use the radio service on a physical Linn device and I believe their software devices. 

As for BubbleUpnp renderers, I'm not sure if such a service is possible? 

Sorry about that. 

1. https://github.com/bazwilliams/openhomedevice#playing-a-track
